Update error handling for typed throws [SE-0413] (#296)
apple/swift-book: TSPL.docc/LanguageGuide/ErrorHandling.md

can be values of any type that conforms to the `Error` protocol.
This approach matches the reality that
you don't know ahead of time every error that could happen
while the code is running,
especially when propagating errors thrown somewhere else.
This approach also reflects the fact that errors can change over time.
New versions of a library can throw new errors ---
including libraries used by your dependencies ---
and the rich complexity of real-world user configurations
can expose failure modes that weren't visible during development or testing.
The error handling code in the examples above
always includes a default case to handle errors
that don't have their own specific `catch` clause.

Most Swift code doesn't specify the type for the errors it throws.
However,
in some special cases,
you might limit code to throwing errors of only one specific type:

- When running code on an embedded system
  that doesn't support dynamic allocation of memory.
  Throwing an instance `any Error` or another boxed protocol type
  requires allocating memory at runtime to store the error.
  Throwing an error of a specific type
  lets Swift allocate that memory upfront instead.

> - When the errors are used only within some unit of code,

Author: This is going to be a bit tricky to put into active voice, and not necessarily clearer — this is a good example of an actor that's poorly defined and unimportant.  It might be better as "When the errors are *contained* within some unit of code"?  Or to rewrite more aggressively:

> When the errors are an implementation detail of some unit of code,
> like a library,
> and aren't part of the interface to that code.

I'll start with that last rewrite, and we can iterate.

>   like a library,

Author: I was going to note that TSPL here follows DevPubs style of generally preferring "like" over "such as" to introduce examples — but doing a search I see the former appears 196 times and the latter appears 133 times.  I think this case reads better with "like" but if you feel strongly (or if there's a style rule I'm overlooking), I'll change it.

>   like a library,

Reviewer: I'm fine with leaving it!

  and aren't part of the interface to that code.
  Because the errors come only from the library,
  and not from other dependencies or the library's clients,
  you can make an exhaustive list of all possible failures.
  And because these errors are an implementation detail of the library,
  and they're always handled within that library.

- In code that only throws errors that were thrown elsewhere,

>   like a function that takes a closure argument

Author: (As line 734 above)

  and propagates any errors from that closure.
  For a comparison between `rethrows`

>   and throwing a specific, generic, error type

Author: Dropped "generic" to make the xref less specific, and easier to phrase.  Started with the half of the comparison that's relevant to the current discussion.

  see <doc:Declarations:Rethrowing-Functions-and-Methods>.

For example,
consider code that summarizes ratings
and uses the following error type:

```swift
enum StatisticsError: Error {
    case noRatings
    case invalidRating(Int)
}
```

To specify that a function throws only `StatisticsError` values as its errors
you write `throws(StatisticsError)` when declaring the function,
instead of just writing `throws`.
This syntax is also called *typed throws*

> because you write the error type after `throws` --- for example:

Author: Rewrote to frame more specifically what the developers should look at the example for.

Swift checks that you don't throw any other errors.
For example,
if you tried to use `VendingMachineError` from examples earlier in this chapter
in the `summarize(_:)` function above,
that code would produce an error at compile time.

You can call a function that uses typed throws
from within a regular throwing function:

```swift
func someThrowingFunction() -> throws {
    let ratings = [1, 2, 3, 2, 2, 1]
    try summarize(ratings)
}
```

The code above doesn't specify an error type for `someThrowingFunction()`,
so it throws `any Error`.
You could also write the error type explicitly as `throws(any Error)` ---
the code below is equivalent to the code above:

```swift
func someThrowingFunction() -> throws(any Error) {
    let ratings = [1, 2, 3, 2, 2, 1]
    try summarize(ratings)
}
```

In this code,
`someThrowingFunction()` propagates any errors that `summarize(_:)` throws.
The errors from `summarize(_:)` are always `StatisticsError` values,
which is also a valid error for `someThrowingFunction()` to throw.
<!-- XXX Expand on subtyping here? -->

In addition to specifying a function's error type,
you can also write a specific error type for a `do`-`catch` statement.

> For example:

Author: Introducing examples this way is pretty typical in TSPL.  It appears in 70 other places — we could start a bulk clean-up pass like https://github.com/apple/swift-book/pull/141 if there's an underlying issue with this phrasing.

```swift
let ratings = []
do throws(StatisticsError) {
    try summarize(ratings)
} catch {
    switch error {
    case .noRatings:
        print("No ratings available")
    case .invalidRating(let rating):
        print("Invalid rating: \(rating)")
    }
}
// Prints "No ratings available"
```

In this code,
writing `do throws(StatisticsError)` indicates that
the `do`-`catch` statement throws `StatisticsError` values as its errors.
Like other `do`-`catch` statements,

> the `catch` clause can either handle every possible error,
> or it can propagate unhandled errors for some surrounding scope to handle.

Reviewer: ```suggestion
the `catch` clause can either handle every possible error
or propagate unhandled errors for some surrounding scope to handle.
```

  ```
-->

## Specifying the Error Type

All of the examples above use the most common kind of error handling,
where the errors that your code throws
can be values of any type that conforms to the `Error` protocol.
This approach matches the reality that
you don't know ahead of time every error that could happen
while the code is running,
especially when propagating errors thrown somewhere else.

> This approach also reflects the fact that errors can change over time.

Reviewer 2: "This approach" starts both this sentence and the prior one. Perhaps use this instead?
```suggestion
It also reflects the fact that errors can change over time.
```

New versions of a library ---
including libraries that your dependencies use ---
can throw new errors,
and the rich complexity of real-world user configurations
can expose failure modes that weren't visible during development or testing.
The error handling code in the examples above
always includes a default case to handle errors
that don't have a specific `catch` clause.

Most Swift code doesn't specify the type for the errors it throws.
However,
you might limit code to throwing errors of only one specific type
in the following special cases:

- When running code on an embedded system
  that doesn't support dynamic allocation of memory.
  Throwing an instance of `any Error` or another boxed protocol type
  requires allocating memory at runtime to store the error.
  In contrast,
  throwing an error of a specific type

>   lets Swift allocate that memory upfront.

Reviewer 2: Using "allocate" again here implies the heap. Perhaps somehing like this:
```suggestion
  lets Swift avoid heap allocation for errors.
```

- When the errors are an implementation detail of some unit of code,
  like a library,
  and aren't part of the interface to that code.
  Because the errors come from only the library,
  and not from other dependencies or the library's clients,
  you can make an exhaustive list of all possible failures.
  And because these errors are an implementation detail of the library,
  they're always handled within that library.

> - In code that throws only errors that were thrown elsewhere,
>   like a function that takes a closure argument
>   and propagates any errors from that closure.

Reviewer 2: I'd like to mention that this is generic code here. We could say "in generic code", or do something like this:

```suggestion
- In code that only propagates errors described by generic parameters, 
  like a function that takes a closure argument
  and propagates any errors from that closure.
```

instead of writing `throw StatisticsError.noRatings`
when throwing an error in a function like this.

When you write a specific error type at the start of the function,
Swift checks that you don't throw any other errors.
For example,
if you tried to use `VendingMachineError` from examples earlier in this chapter
in the `summarize(_:)` function above,
that code would produce an error at compile time.

You can call a function that uses typed throws
from within a regular throwing function:

```swift
func someThrowingFunction() -> throws {
    let ratings = [1, 2, 3, 2, 2, 1]
    try summarize(ratings)
}
```

The code above doesn't specify an error type for `someThrowingFunction()`,
so it throws `any Error`.
You could also write the error type explicitly as `throws(any Error)`;
the code below is equivalent to the code above:

```swift
func someThrowingFunction() -> throws(any Error) {
    let ratings = [1, 2, 3, 2, 2, 1]
    try summarize(ratings)
}
```

In this code,
`someThrowingFunction()` propagates any errors that `summarize(_:)` throws.
The errors from `summarize(_:)` are always `StatisticsError` values,
which is also a valid error for `someThrowingFunction()` to throw.

> <!-- XXX Expand on subtyping here? -->

Reviewer 2: I don't think you need to expand on subtyping. However, I would like to describe the `Never` case, because it's an important part of understanding how typed throws fits in the type system. For example, it could be something like this:

> Just like you can write a function that never returns with a return type of `Never`, you can write a function that never throws with `throws(Never)`:

    ```swift
    func nonThrowingFunction() throws(Never) {
      // ...
    }
    ```

> This function cannot throw because it is impossible to create a value of type `Never` to throw.

In addition to specifying a function's error type,
you can also write a specific error type for a `do`-`catch` statement.
For example:

```swift
let ratings = []
do throws(StatisticsError) {
    try summarize(ratings)
} catch {
    switch error {
    case .noRatings:
        print("No ratings available")
    case .invalidRating(let rating):
        print("Invalid rating: \(rating)")
    }
}
// Prints "No ratings available"
```

In this code,
writing `do throws(StatisticsError)` indicates that
the `do`-`catch` statement throws `StatisticsError` values as its errors.
Like other `do`-`catch` statements,
the `catch` clause can either handle every possible error
or propagate unhandled errors for some surrounding scope to handle.
This code handles all of the errors,
using a `switch` statement with one case for each enumeration value.
Like other `catch` clauses that don't have a pattern,
the clause matches any error
and binds the error to a local constant named `error`.
Because the `do`-`catch` statement throws `StatisticsError` values,
`error` is a value of type `StatisticsError`.

> <!-- XXX show multiple catch clauses with different patterns? -->

Reviewer 2: I don't think you have to show multiple catch clauses specifically. It falls out from other sections.
